src/parser/jobs/blm/Gauge.js

```diff
@@ -7,210 +7,291 @@ import ACTIONS from 'data/ACTIONS'
+			this._umbralIceStacks = Math.min(this._umbralIceStacks + 1, MAX_ASTRAL_UMBRAL_STACKS)
+			this.addEvent()
 		}
-		//getting rid of UHs one AF action at a time
-		if (AF_ACTIONS.includes(abilityId) && this._UH > 0 && this._UI === 0) {
-			this._UH --
-			this._UH = Math.max(this._UH, 0)
+	}
+
+	onGainAstralFireStack(event) {
+		if (this._umbralIceStacks > 0) {
+			this.onAstralUmbralTimeout()
+		} else {
+			this._astralUmbralStackTimer = event.timestamp
+			this._astralFireStacks = Math.min(this._astralFireStacks + 1, MAX_ASTRAL_UMBRAL_STACKS)
+			this.addEvent()
 		}
+	}
+
+	onGainMaxAstralFireStacks(event) {
+		this._umbralIceStacks = 0
+		this._astralFireStacks = MAX_ASTRAL_UMBRAL_STACKS
+		this._astralUmbralStackTimer = event.timestamp
+		this.addEvent()
+	}
 
-		//Flare stuff
-		if (abilityId === ACTIONS.FLARE.id) {
-			this._AF = 3
-			this._UI = 0
-			this._AFUITimer = event.timestamp
-			if (this._UH > 0) {
-				this._UH = 0
+	onGainMaxUmbralIceStacks(event) {
+		this._astralFireStacks = 0
+		this._umbralIceStacks = MAX_ASTRAL_UMBRAL_STACKS
+		this._astralUmbralStackTimer = event.timestamp
+		this.addEvent()
+	}
+
+	onTransposeStacks(event) {
+		if (this._astralFireStacks > 0 || this._umbralIceStacks > 0) {
+			this._astralUmbralStackTimer = event.timestamp
+			if (this._astralFireStacks > 0) {
+				this._astralFireStacks = 0
+				this._umbralIceStacks = 1
+			} else {
+				this._astralFireStacks = 1
+				this._umbralIceStacks = 0
 			}
+			this.addEvent()
 		}
+	}
 
-		//do F3 things
-		if (abilityId === ACTIONS.FIRE_III.id) {
-			this._UI = 0
-			this._AF = 3
-			this._AFUITimer = event.timestamp
+	tryConsumeUmbralHearts(event, count, force = false) {
+		if (this._umbralHeartStacks > 0 && (this._astralFireStacks > 0 || force)) {
+			this._umbralHeartStacks = Math.max(this._umbralHeartStacks - count, 0)
+			this.addEvent()
 		}
+	}
 
-		//do B3 things
-		if (abilityId === ACTIONS.BLIZZARD_III.id) {
-			this._UI = 3
-			this._AF = 0
-			this._AFUITimer = event.timestamp
+	updateStackTimers(event) {
+		if ((this._astralFireStacks > 0 || this._umbralIceStacks > 0) &&
+			(event.timestamp - this._astralUmbralStackTimer > ASTRAL_UMBRAL_DURATION)
+		) {
+			this.onAstralUmbralTimeout()
 		}
 
-		//Foul support and poly adjustment
-		if (abilityId === ACTIONS.FOUL.id) {
-			this._poly = 0
+		if (this._hasEnochian) {
+			const enoRunTime = event.timestamp - this._enochianTimer
+			if (enoRunTime >= ENOCHIAN_DURATION_REQUIRED) {
+				this._enochianTimer = event.timestamp - (enoRunTime - ENOCHIAN_DURATION_REQUIRED)
+				this.onGainPolyglot()
+			}
 		}
+	}
 
-		if (abilityId === ACTIONS.TRANSPOSE.id) {
-			if (this._AF > 0) {
-				this._AF = 0
-				this._UI = 1
-				this._AFUITimer = event.timestamp
-			}
-			if (this._UI > 0) {
-				this._UI = 0
-				this._AF = 1
-				this._AFUITimer = event.timestamp
+	_onCast(event) {
+		const abilityId = event.ability.guid
+
+		switch (abilityId) {
+		case ACTIONS.ENOCHIAN.id:
```

Comment:
Holy switch batman.

w/o the normalise, this could probably be done with a bunch of hooks, though I don't particularly mind either way.